### PR TITLE
Allow updates to Logstorage resource limits

### PR DESCRIPTION
**Pull request by Suraiya-Hameed**
## Description
Updating the Logstroage resource limit fails as storage requirements are immutable from ECK 1.0 https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-orchestration.html#k8s-orchestration-limitations

This PR uses hash of the volumeClaimTemplates as NodeSet name so a new StatefulSet is created if storage requirements change.

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/render/logstorage.go [hunk 1]
```diff
@@ -522,22 +526,32 @@ func (es elasticsearchComponent) elasticsearchCluster() *esv1.Elasticsearch {
 			NodeSets: []esv1.NodeSet{
 				{
 					Count: int32(nodeConfig.Count),
-					Name:  "es",
+					Name: "es-" + nodeSetNameSuffix(pvcTemplate),
```
Comment (rene-dekker) [inline pkg/render/logstorage.go:529]: You can omit the "es-" bit

File: pkg/render/logstorage_test.go [hunk 2]
```diff
@@ -267,6 +267,99 @@ var _ = Describe("Elasticsearch rendering tests", func() {
 
 		Context("Deleting LogStorage", deleteLogStorageTests(operatorv1.ClusterManagementTypeStandalone))
 
+		Context("Updating LogStorage resource", func() {
```
Comment (rene-dekker) [inline pkg/render/logstorage_test.go:270]: Does this test fail without your fix? I'm not sure if the fake k8s client we use actually captures the problem.

Comment (rene-dekker) [reply pkg/render/logstorage_test.go:270]: Could you add an assertion on a spec change?

File: pkg/render/logstorage.go (hunk 1 above)
Comment (Brian-McM) [reply pkg/render/logstorage.go:529]: Is a prefix added to the node set automatically? If not we should add "tigera" or something like that.

File: pkg/render/logstorage.go [hunk 3]
```diff
@@ -522,22 +526,32 @@ func (es elasticsearchComponent) elasticsearchCluster() *esv1.Elasticsearch {
...
 }
 
+// nodeSetNameSuffix returns thumbprint of PersistentVolumeClaim object as string.
+// As storage requirements of NodeSets are immutable,
+// renaming a NodeSet automatically creates a new StatefulSet with new PersistentVolumeClaim.
+func nodeSetNameSuffix(pvcTemplate corev1.PersistentVolumeClaim) string{
+	pvcTemplateHash := md5.New()
+	templateBytes, _ := json.Marshal(pvcTemplate)
+	pvcTemplateHash.Write(templateBytes)
+	// Slice to fit NodeSets Name's max length of 23 char
```
Comment (stevegaossou) [inline pkg/render/logstorage.go:552]: @Suraiya-Hameed could you reference the actual Elastic doc that states this? 

Important restriction and good to provide a reference here.

File: pkg/render/logstorage.go (hunk 1 above)
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:529]: no, what we give is used as it is for NodeSet name, with the current change pod name becomes something like `tigera-secure-es-es-1dacabc7d15654312c8c-0` from original `tigera-secure-es-es-0`.

Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:529]: I'll remove `es-` prefix from the NodeSet name, it'll be less redundant, so the statefulset generated will be `tigera-secure-es-<thumbprint>`

Comment (rene-dekker) [reply pkg/render/logstorage.go:529]: Is it hard to get a shorter hash?

Comment (stevegaossou) [reply pkg/render/logstorage.go:529]: > You can omit the "es-" bit
> Is a prefix added to the node set automatically? If not we should add "tigera" or something like that.

Agree with Rene. You don't need any prefix in the NodeSet name, because ECK already adds another prefix (from the Elasticsearch resource name). 

I was curious about this myself, so I looked at the Elastic docs. 

[Doc states](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-orchestration.html#k8s-statefulsets): 
> the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name. Each Pod will be assigned the StatefulSet name suffixed by an ordinal. The corresponding Elasticsearch node has the same name as the Pod.

So for example, if the NodeSet name is `es` and our Elasticsearch resource name is `tigera-secure` (which is what it is currently), then from the Elastic docs it sounds like StatefulSet name should be `tigera-secure-es`. It's actually `tigera-secure-es-es` (I checked in a running cluster). 

The Elasticsearch pod created from the StatefulSet is called `tigera-secure-es-es-0`. 

So other than the double `es` it seems to fit what the documentation says.

File: pkg/render/logstorage.go (hunk 3 above)
Comment (stevegaossou) [reply pkg/render/logstorage.go:552]: I couldn't find a mention of this max length in the docs. Could you share the doc you found on this?

File: pkg/render/logstorage.go [hunk 4]
```diff
@@ -522,22 +526,33 @@ func (es elasticsearchComponent) elasticsearchCluster() *esv1.Elasticsearch {
...
+// nodeSetName returns thumbprint of PersistentVolumeClaim object as string.
+// As storage requirements of NodeSets are immutable,
+// renaming a NodeSet automatically creates a new StatefulSet with new PersistentVolumeClaim.
+// https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-orchestration.html#k8s-orchestration-limitations
+func nodeSetName(pvcTemplate corev1.PersistentVolumeClaim) string{
+	pvcTemplateHash := md5.New()
+	templateBytes, _ := json.Marshal(pvcTemplate)
+	pvcTemplateHash.Write(templateBytes)
+	// Slice to fit NodeSets Name's max length of 23 char
+	return hex.EncodeToString(pvcTemplateHash.Sum(nil))[:23]
```
Comment (stevegaossou) [inline pkg/render/logstorage.go:554]: Everything hinges on this returning a unique hash. What guarantees the first 20 chars are unique?

File: pkg/render/logstorage_test.go [hunk 5]
```diff
@@ -267,6 +267,104 @@ var _ = Describe("Elasticsearch rendering tests", func() {
...
+			It("should create new NodeSet", func() {
+				logStorage2 := &operator.LogStorage{
```
Comment (stevegaossou) [inline pkg/render/logstorage_test.go:272]: Minor: Why is this called `logStroage2`? The number 2 seems like noise (not important).

File: pkg/render/logstorage.go (hunk 3 above)
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:553]: https://github.com/elastic/cloud-on-k8s/blob/0404e804124f51fb7095f2ff2eb905b399067b42/pkg/apis/elasticsearch/v1beta1/elasticsearch_types.go#L63

File: pkg/render/logstorage_test.go (hunk 2 above)
Comment (stevegaossou) [reply pkg/render/logstorage_test.go:270]: Everything about this fix hinges on the function `nodeSetName()` generating a unique (20-char) hash substring given corev1.PersistentVolumeClaim instances that contain different values. 

Would like to see a few more tests perhaps just for `nodeSetName()` itself (unit style tests) to illustrate this. Are there an edge cases where this isn't true?

File: pkg/render/logstorage.go (hunk 1 above)
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go]: >the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name. Each Pod will be assigned the StatefulSet name suffixed by an ordinal. The corresponding Elasticsearch node has the same name as the Pod.
https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-orchestration.html#k8s-statefulsets
Thats how we are getting `tigera-secure-es-es-0`

File: pkg/render/logstorage.go (hunk 3 above)
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:553]: I couldn't find documentation around it, I got the error saying max length is 23 when I used full hash output (32 char).

File: pkg/render/logstorage.go (hunk 4 above)
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:554]: 128bit is the minimum length of hash output, that would be 32 char. I was looking at options to get 23 char unique value, but all the docs I referred suggested truncating as collision will be pretty low even with that.

Comment (Brian-McM) [reply pkg/render/logstorage.go:554]: > Everything hinges on this returning a unique hash. What guarantees the first 20 chars are unique?
It would be _extremely_ unlikely that even the first 5 characters would be the same, using 23 characters is quite the overkill. One character differences in two strings going through a hash function usually generate wildly different hashes.

This will generate huge pod names, which might seem a little weird to our customers, and I think the likely hood of collision with even 5 suffixes is so small.

One other note is that only one suffix ever exists at a time, and we don't need to generate a hash that's never been used before, we only need to generate a hash that was different from the previous ones. The likely hood of generating two hash in a row with 5 character likes that have different generated from unique PVCs is extremely unlikely, to say the least.

Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go]: I've switched to non-cryptographic hash with smaller output length.

File: pkg/render/logstorage.go [hunk 6]
```diff
@@ -522,22 +526,32 @@ func (es elasticsearchComponent) elasticsearchCluster() *esv1.Elasticsearch {
...
 	}
 }
 
+// nodeSetName returns thumbprint of PersistentVolumeClaim object as string.
+// As storage requirements of NodeSets are immutable,
+// renaming a NodeSet automatically creates a new StatefulSet with new PersistentVolumeClaim.
+// https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-orchestration.html#k8s-orchestration-limitations
+func nodeSetName(pvcTemplate corev1.PersistentVolumeClaim) string{
+	pvcTemplateHash := fnv.New64a()
+	templateBytes, _ := json.Marshal(pvcTemplate)
```
Comment (Brian-McM) [inline pkg/render/logstorage.go:551]: This returns an error and it should be handled. I think the pattern in render is to panic, since we don't return errors. If there's an error the below line likely panicsince templateBytes will be nil.

Comment (Brian-McM) [reply pkg/render/logstorage.go:551]: Maybe confirm with @caseydavenport or @tmjd that this is OK, sometimes our opinions with panicing differ. This needs to be handle somehow though.

Comment (caseydavenport) [reply pkg/render/logstorage.go]: blech, yeah it's probably fine.

I don't really like the pattern that we've established to "panic" in render - it's not very graceful or resilient to potential bugs in the code.

Ideally we'd propagate an error back up to report to the user in the TigeraStatus, but that's a larger change than I want to make you do in this PR.

Something better but not as good would be to log out a warning and then return sensible output. e.g., if we can't calculate a hash, we just return `es` like we do today?

Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go]: >Something better but not as good would be to log out a warning and then return sensible output. e.g., if we can't calculate a hash, we just return es like we do today?

Sure sounds better than panicking.
